File: node-ts/api-exemplo/src/controllers/user.controller.ts
import type { Request, Response } from 'express';
import * as userModel from '../models/user.model.js';

// "Controller" = recebe a requisição, chama o Model, devolve a resposta.
// Equivalente à view do Django ou ao Controller do Laravel.

export function index(req: Request, res: Response) {
  res.json(userModel.findAll());
}

export function show(req: Request, res: Response) {
  const id = Number(req.params.id);
  const user = userModel.findById(id);

  if (!user) {
    // debug tip: coloque um console.log(id) aqui se quiser ver o
    // valor chegando do request.params
    return res.status(404).json({ error: 'usuário não encontrado' });
  }

  res.json(user);
}

export function store(req: Request, res: Response) {
  const { name, email } = req.body;

  if (!name || !email) {
    return res.status(400).json({ error: 'name e email são obrigatórios' });
  }

  const user = userModel.create({ name, email });
  res.status(201).json(user);
}

File: golang/api-exemplo/controllers/user_controller.go
package controllers

import (
	"encoding/json"
	"net/http"

	"aprendizado/api-exemplo/models"
)

// "Controller" = recebe a requisição, chama o Model, escreve a resposta.

func Index(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(models.FindAll())
}

func Show(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	user, ok := models.FindByID(id)

	if !ok {
		// debug tip: um fmt.Println(id) aqui mostra o valor cru vindo da URL
		http.Error(w, `{"error":"usuário não encontrado"}`, http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(user)
}

type storeInput struct {
	Name  string `json:"name"`
	Email string `json:"email"`
}

func Store(w http.ResponseWriter, r *http.Request) {
	var input storeInput
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		http.Error(w, `{"error":"body inválido"}`, http.StatusBadRequest)
		return
	}

	user, err := models.Create(input.Name, input.Email)
	if err != nil {
		http.Error(w, `{"error":"`+err.Error()+`"}`, http.StatusBadRequest)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(user)
}
